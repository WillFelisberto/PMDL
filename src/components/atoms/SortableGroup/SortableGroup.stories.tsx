import type { Meta, StoryObj } from '@storybook/react';

import { SortableGroup } from './SortableGroup';

const meta: Meta<typeof SortableGroup> = {
  title: 'Atoms/SortableGroup',
  component: SortableGroup,
  tags: ['autodocs'],
  args: {
    children: 'Example',
    variant: 'default'
  }
};

export default meta;

type Story = StoryObj<typeof SortableGroup>;

export const Default: Story = {};
