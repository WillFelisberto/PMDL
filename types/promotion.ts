export type ConditionType =
  | 'comparator'
  | 'iterator'
  | 'group-iterator'
  | 'quantifier'
  | 'operator'
  | 'operation';

export type ConditionBase = {
  id: string;
  type: ConditionType;
  attributes?: Array<{ name: string; value: string }>;
};

export type ComparatorCondition = ConditionBase & {
  type: 'comparator';
  comparatorType:
    | 'equals'
    | 'not-equals'
    | 'less-than'
    | 'less-than-or-equals'
    | 'greater-than'
    | 'greater-than-or-equals'
    | 'contains'
    | 'starts-with'
    | 'ends-with'
    | 'isoneof'
    | 'isnotoneof'
    | 'includes'
    | 'includes-any'
    | 'includes-all';
  leftValue: string;
  rightValue: string;
  dataType: string;
};

export type ValueType = {
  type: 'value' | 'constant' | 'array' | 'null';
  dataType?: string;
  value?: string | number | boolean;
  values?: ValueType[];
};
export type IteratorCondition = ConditionBase & {
  type: 'iterator';
  iteratorType: 'next' | 'every' | 'up-to-and-including' | 'custom';
  collectionName: string;
  elementName: string;
  number?: number;
  sortBy?: string;
  sortOrder?: 'asc' | 'desc';
  condition?: PromotionCondition;
};

export type GroupIteratorCondition = ConditionBase & {
  type: 'group-iterator';
  iteratorType: 'every-group' | 'up-to-and-including-group' | 'custom';
  number?: number;
  conditions?: PromotionCondition[];
};

export type QuantifierCondition = ConditionBase & {
  type: 'quantifier';
  quantifierType: 'exactly' | 'at-least' | 'at-most' | 'all' | 'custom';
  number?: number;
  collectionName: string;
  elementName: string;
  condition?: PromotionCondition;
  customType?: string;
};

export type OperatorCondition = ConditionBase & {
  type: 'operator';
  operatorType: 'and' | 'or' | 'not' | 'custom';
  conditions: PromotionCondition[];
  customType?: string; // Para operadores personalizados
};

export type OperationCondition = ConditionBase & {
  type: 'operation';
  operationType: 'union' | 'anded-union' | 'custom';
  iterators: PromotionCondition[];
  customType?: string;
};
export type PromotionCondition =
  | ComparatorCondition
  | IteratorCondition
  | GroupIteratorCondition
  | QuantifierCondition
  | OperatorCondition
  | OperationCondition;

export type DiscountStructure = {
  calculatorType: string;
  discountType: 'percentOff' | 'amountOff' | 'fixedPrice';
  adjuster: number;
  target?: 'order' | 'shipping' | 'item';
};

export type Offer = {
  discountStructures: DiscountStructure[];
};

export type Promotion = {
  displayName: string;
  description: string;
  enabled: boolean;
  priority: number;
  qualifier: PromotionCondition | undefined;
  offer: Offer;
  conditions: PromotionCondition[];
  sites: Array<{ repositoryId: string }>;
};
