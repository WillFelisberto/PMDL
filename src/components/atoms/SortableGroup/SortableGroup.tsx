'use client';
import { OperatorCondition, PromotionCondition } from '@/types/promotion';
import {
  DndContext,
  closestCenter,
  KeyboardSensor,
  PointerSensor,
  useSensor,
  useSensors
} from '@dnd-kit/core';
import { useSortable } from '@dnd-kit/sortable';
import { arrayMove, SortableContext, verticalListSortingStrategy } from '@dnd-kit/sortable';
import { CSS } from '@dnd-kit/utilities';
import { GripVertical, X, Plus } from 'lucide-react';
import { useState } from 'react';

import { SortableCondition } from '../SortableCondition';

export const SortableGroup = ({
  group,
  onUpdate,
  onDelete,
  depth = 0
}: {
  group: OperatorCondition;
  onUpdate: (updated: OperatorCondition) => void;
  onDelete: () => void;
  depth?: number;
}) => {
  const [isHovered, setIsHovered] = useState(false);
  const { attributes, listeners, setNodeRef, transform, transition } = useSortable({
    id: group.id
  });
  const sensors = useSensors(useSensor(PointerSensor), useSensor(KeyboardSensor));

  const style = {
    transform: CSS.Transform.toString(transform),
    transition,
    marginLeft: `${depth * 24}px` // Indentação para grupos aninhados
  };

  const handleDragEnd = (event: any) => {
    const { active, over } = event;
    if (active.id !== over.id) {
      onUpdate({
        ...group,
        conditions: arrayMove(
          group.conditions,
          group.conditions.findIndex((c) => c.id === active.id),
          group.conditions.findIndex((c) => c.id === over.id)
        )
      });
    }
  };

  return (
    <div
      ref={setNodeRef}
      style={style}
      className="relative bg-blue-50 rounded-lg border-2 border-blue-200 p-4 group"
      onMouseEnter={() => setIsHovered(true)}
      data-testid="sortable-group"
      onMouseLeave={() => setIsHovered(false)}
    >
      {/* Header do Grupo */}
      <div className="flex items-center gap-2 mb-2">
        <button
          {...listeners}
          {...attributes}
          className="cursor-grab text-blue-500 hover:text-blue-700"
        >
          <GripVertical size={16} />
        </button>

        <select
          className="px-2 py-1 rounded-md bg-white border"
          value={group.operatorType}
          onChange={(e) =>
            onUpdate({
              ...group,
              operatorType: e.target.value as 'and' | 'or' | 'not' | 'custom'
            })
          }
        >
          <option value="and">AND</option>
          <option value="or">OR</option>
          <option value="not">NOT</option>
        </select>

        {isHovered && (
          <button
            onClick={onDelete}
            data-testid="delete-button"
            className="ml-auto text-red-500 hover:text-red-700"
          >
            <X size={16} />
          </button>
        )}
      </div>

      {/* Condições Filhas */}
      <DndContext sensors={sensors} collisionDetection={closestCenter} onDragEnd={handleDragEnd}>
        <SortableContext items={group.conditions} strategy={verticalListSortingStrategy}>
          <div className="space-y-2">
            {group.conditions.map((item) =>
              'operatorType' in item ? (
                <SortableGroup
                  key={item.id}
                  group={item}
                  onUpdate={(updated) => {
                    const newConditions = group.conditions.map((c) =>
                      c.id === updated.id ? updated : c
                    );
                    onUpdate({ ...group, conditions: newConditions });
                  }}
                  onDelete={() => {
                    const newConditions = group.conditions.filter((c) => c.id !== item.id);
                    onUpdate({ ...group, conditions: newConditions });
                  }}
                  depth={depth + 1}
                />
              ) : (
                <SortableCondition
                  key={item.id}
                  condition={item}
                  onUpdate={(updated) => {
                    const newConditions = group.conditions.map((c) =>
                      c.id === updated.id ? updated : c
                    );
                    onUpdate({ ...group, conditions: newConditions });
                  }}
                  onDelete={() => {
                    const newConditions = group.conditions.filter((c) => c.id !== item.id);
                    onUpdate({ ...group, conditions: newConditions });
                  }}
                />
              )
            )}
          </div>
        </SortableContext>
      </DndContext>

      {/* Botões de Ação */}
      <div className="mt-2 flex gap-2">
        <button
          onClick={() => {
            const newCondition: PromotionCondition = {
              id: crypto.randomUUID(),
              type: 'comparator',
              comparatorType: 'equals',
              leftValue: '',
              rightValue: '',
              dataType: 'java.lang.String'
            };
            onUpdate({
              ...group,
              conditions: [...group.conditions, newCondition]
            });
          }}
          data-testid="add-condition-button"
          className="flex items-center gap-1 text-sm text-blue-500 hover:text-blue-700"
        >
          <Plus size={14} /> Condição
        </button>

        <button
          onClick={() => {
            const newGroup: OperatorCondition = {
              id: crypto.randomUUID(),
              type: 'operator',
              operatorType: 'and',
              conditions: []
            };
            onUpdate({ ...group, conditions: [...group.conditions, newGroup] });
          }}
          data-testid="add-group-button"
          className="flex items-center gap-1 text-sm text-blue-500 hover:text-blue-700"
        >
          <Plus size={14} /> Grupo
        </button>
      </div>
    </div>
  );
};
