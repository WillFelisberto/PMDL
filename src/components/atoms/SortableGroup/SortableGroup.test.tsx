import { SortableGroup } from './SortableGroup';

import { render, screen, fireEvent } from '@/tests/test-utils';
import { OperatorCondition } from '@/types/promotion';

describe('SortableGroup', () => {
  let mockGroup: OperatorCondition;
  let mockOnUpdate: jest.Mock;
  let mockOnDelete: jest.Mock;

  beforeAll(() => {
    Object.defineProperty(global, 'crypto', {
      value: {
        randomUUID: jest.fn(() => 'mocked-uuid')
      }
    });
  });

  beforeEach(() => {
    mockGroup = {
      id: 'group-1',
      type: 'operator',
      operatorType: 'and',
      conditions: []
    };

    mockOnUpdate = jest.fn();
    mockOnDelete = jest.fn();
  });

  it('deve renderizar corretamente', () => {
    const { container } = render(
      <SortableGroup group={mockGroup} onUpdate={mockOnUpdate} onDelete={mockOnDelete} />
    );

    expect(screen.getByRole('button', { name: /condição/i })).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /grupo/i })).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('deve permitir alterar o operador lógico', () => {
    render(<SortableGroup group={mockGroup} onUpdate={mockOnUpdate} onDelete={mockOnDelete} />);

    const select = screen.getByRole('combobox');
    fireEvent.change(select, { target: { value: 'or' } });

    expect(mockOnUpdate).toHaveBeenCalledWith({
      ...mockGroup,
      operatorType: 'or'
    });
  });

  it('deve chamar onDelete corretamente ao remover o grupo', () => {
    render(<SortableGroup group={mockGroup} onUpdate={mockOnUpdate} onDelete={mockOnDelete} />);

    // Simula hover para o botão de delete aparecer
    const groupElement = screen.getByTestId('sortable-group');
    fireEvent.mouseEnter(groupElement);

    const deleteButton = screen.getByTestId('delete-button');
    fireEvent.click(deleteButton);

    expect(mockOnDelete).toHaveBeenCalled();
  });

  it('deve adicionar uma nova condição ao grupo', () => {
    render(<SortableGroup group={mockGroup} onUpdate={mockOnUpdate} onDelete={mockOnDelete} />);

    const addConditionButton = screen.getByTestId('add-condition-button');
    fireEvent.click(addConditionButton);

    expect(mockOnUpdate).toHaveBeenCalledWith(
      expect.objectContaining({
        conditions: expect.arrayContaining([expect.objectContaining({ type: 'comparator' })])
      })
    );
  });

  it('deve adicionar um novo grupo dentro do grupo', () => {
    render(<SortableGroup group={mockGroup} onUpdate={mockOnUpdate} onDelete={mockOnDelete} />);

    const addGroupButton = screen.getByTestId('add-group-button');
    fireEvent.click(addGroupButton);

    expect(mockOnUpdate).toHaveBeenCalledWith(
      expect.objectContaining({
        conditions: expect.arrayContaining([expect.objectContaining({ type: 'operator' })])
      })
    );
  });
});
