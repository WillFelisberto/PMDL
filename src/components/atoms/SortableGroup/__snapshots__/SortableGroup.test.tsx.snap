// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortableGroup deve renderizar corretamente 1`] = `
<div>
  <div
    class="relative bg-blue-50 rounded-lg border-2 border-blue-200 p-4 group"
    data-testid="sortable-group"
    style="margin-left: 0px;"
  >
    <div
      class="flex items-center gap-2 mb-2"
    >
      <button
        aria-describedby=""
        aria-disabled="false"
        aria-roledescription="sortable"
        class="cursor-grab text-blue-500 hover:text-blue-700"
        role="button"
        tabindex="0"
      >
        GripVertical Icon
      </button>
      <select
        class="px-2 py-1 rounded-md bg-white border"
      >
        <option
          value="and"
        >
          AND
        </option>
        <option
          value="or"
        >
          OR
        </option>
        <option
          value="not"
        >
          NOT
        </option>
      </select>
    </div>
    <div
      class="space-y-2"
    />
    <div
      id="DndDescribedBy-0"
      style="display: none;"
    >
      
    To pick up a draggable item, press the space bar.
    While dragging, use the arrow keys to move the item.
    Press space again to drop the item in its new position, or press escape to cancel.
  
    </div>
    <div
      aria-atomic="true"
      aria-live="assertive"
      id="DndLiveRegion-0"
      role="status"
      style="position: fixed; top: 0px; left: 0px; width: 1px; height: 1px; margin: -1px; border: 0px; padding: 0px; overflow: hidden; clip-path: inset(100%); white-space: nowrap;"
    />
    <div
      class="mt-2 flex gap-2"
    >
      <button
        class="flex items-center gap-1 text-sm text-blue-500 hover:text-blue-700"
        data-testid="add-condition-button"
      >
        Plus Icon
         Condição
      </button>
      <button
        class="flex items-center gap-1 text-sm text-blue-500 hover:text-blue-700"
        data-testid="add-group-button"
      >
        Plus Icon
         Grupo
      </button>
    </div>
  </div>
</div>
`;
